keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -2689,90 +2813,72 @@ previous answers based on experience in the field.
 
 ###### How can an operator determine if the feature is in use by workloads?
 
-<!--
-Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
-checking if there are objects with field X set) may be a last resort. Avoid
-logs or events for this purpose.
--->
+There will be pods which have a non-empty PodSpec.ResourceClaims field and ResourceClaim objects.
 
-Metrics in kube-scheduler (names to be decided):
-- number of classes using structured parameters
-- number of claims which currently are allocated with structured parameters
+Metrics in kube-controller-manager about total and allocated ResourceClaims.
 
 ###### How can someone using this feature know that it is working for their instance?
 
-<!--
-For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
-for each individual pod.
-Pick one more of these and delete the rest.
-Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
-and operation of this feature.
-Recall that end users cannot usually observe component logs or access metrics.
--->
-
 - [X] API .status
   - Other field: ".status.allocation" will be set for a claim using structured parameters
     when needed by a pod.
 
 ###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
 
-<!--
-This is your opportunity to define what "normal" quality of service looks like
-for a feature.
-
-It's impossible to provide comprehensive guidance, but at the very
-high level (needs more precise definitions) those may be things like:
-  - per-day percentage of API calls finishing with 5XX errors <= 1%
-  - 99% percentile over day of absolute value from (job creation time minus expected
-    job creation time) for cron job <= 10%
-  - 99.9% of /health requests per day finish with 200 code
-
-These goals will help you determine what you need to measure (SLIs) in the next
-question.
--->
+For Pods not using ResourceClaims, the same SLOs apply as before.
```

Comment:
Example: Pods with resource claims getting scheduled, but then failing to obtain the resources should almost never happen. Some SLO around that would be useful.

Step through the process and see where failures might happen, and the ones that are not user error should be tracked with a metric and have an SLO, if it's practical to do so. For example, pods failing to schedule is not useful, because it just may be because there is no capacity. But pods getting scheduled, then failing to actually get the resources because of DRA or runtime failures indicate a problem and should be surfaced via a metric and have an SLO.

Reply:
We have a unique issue of someone using the DRA API, but only Device Plugin is enabled. How can we detect that?

Reply:
> "Obtaining resources" is something that happens inside the DRA driver. The kubelet only sees "NodePrepareResources fails", but doesn't get any further details because it wouldn't make a difference: all it can do is retry.

So, watching for NodePrepareResources fails would be one thing; in proportion to total pod failures even more so. We should already have metrics for those, we can just document it here.

> Enabled in the kubelet?

No, I meant more like, the user specifies some DRA resource claims, but there is no driver anywhere in the cluster registered. I guess there would just be no resource slices and the pod would fail to schedule, and they would see that in the pod status / events. That should be good enough.

Reply:
> So, watching for NodePrepareResources fails would be one thing

I added one paragraph about that and monitoring it with `dra_operations_duration`.

> No, I meant more like, the user specifies some DRA resource claims, but there is no driver anywhere in the cluster registered. ... pod would fail to schedule

Does this really belong under "What are the reasonable SLOs (Service Level Objectives) for the enhancement?"

It feels more like a failure scenario, except that it's not really a failure - DRA works as it should! :sweat_smile:

Reply:
> Does this really belong under "What are the reasonable SLOs (Service Level Objectives) for the enhancement?"
> 
> It feels more like a failure scenario, except that it's not really a failure - DRA works as it should! 😅

Agreed - we don't need anything on this :)